arrow-flight/tests/encode_decode.rs

```diff
@@ -0,0 +1,408 @@
+        decoded_data[1].payload,
+        DecodedPayload::RecordBatch(_)
+    ));
+    assert!(matches!(decoded_data[2].payload, DecodedPayload::Schema(_)));
+    assert!(matches!(
+        decoded_data[3].payload,
+        DecodedPayload::RecordBatch(_)
+    ));
+}
+
+#[tokio::test]
+async fn test_mismatched_schema_message() {
+    // Model sending schema that is mismatched with the data
+    // and expect an error
+    async fn do_test(batch1: RecordBatch, batch2: RecordBatch, expected: &str) {
+        let encode_stream1 = FlightDataEncoderBuilder::default()
+            .build(futures::stream::iter(vec![Ok(batch1.clone())]))
+            // take only schema message from first stream
+            .take(1);
+        let encode_stream2 = FlightDataEncoderBuilder::default()
+            .build(futures::stream::iter(vec![Ok(batch2.clone())]))
+            // take only data message from second
+            .skip(1);
+
+        // append the two streams
+        let encode_stream = encode_stream1.chain(encode_stream2);
+
+        // FlightRecordBatchStream errors if the schema changes
+        let decode_stream = FlightRecordBatchStream::new_from_flight_data(encode_stream);
+        let result: Result<Vec<_>, FlightError> = decode_stream.try_collect().await;
+
+        let err = result.unwrap_err().to_string();
+        assert!(
+            err.contains(expected),
+            "could not find '{expected}' in '{err}'"
+        );
+    }
+
+    // primitive batch first (has more columns)
+    do_test(
+        make_primative_batch(5),
+        make_dictionary_batch(3),
+        "Error decoding ipc RecordBatch: Io error: Invalid data for schema",
+    )
+    .await;
+
+    // dictioanry batch first
+    do_test(
+        make_dictionary_batch(3),
+        make_primative_batch(5),
+        "Error decoding ipc RecordBatch: Invalid argument error",
+    )
+    .await;
+}
+
+/// Make a primtive batch for testing
+///
+/// Example:
+/// i: 0, 1, None, 3, 4
+/// f: 5.0, 4.0, None, 2.0, 1.0
+fn make_primative_batch(num_rows: usize) -> RecordBatch {
+    let i: UInt8Array = (0..num_rows)
+        .map(|i| {
+            if i == num_rows / 2 {
+                None
+            } else {
+                Some(i.try_into().unwrap())
+            }
+        })
+        .collect();
+
+    let f: Float64Array = (0..num_rows)
+        .map(|i| {
+            if i == num_rows / 2 {
+                None
+            } else {
+                Some((num_rows - i) as f64)
+            }
+        })
+        .collect();
+
+    RecordBatch::try_from_iter(vec![("i", Arc::new(i) as ArrayRef), ("f", Arc::new(f))])
+        .unwrap()
+}
+
+fn make_dictionary_batch(num_rows: usize) -> RecordBatch {
+    let values: Vec<_> = (0..num_rows)
+        .map(|i| {
+            if i == i / 2 {
+                None
+            } else {
+                // repeat some values for low cardinality
+                let v = i / 3;
+                Some(format!("value{v}"))
+            }
```

Comment:
```suggestion
            // repeat some values for low cardinality
            (i != i / 2).then(|| format!("value{}", i / 3))
```
I also wonder if `i == i / 2` is what you meant to write, as that will only be true for `0`

Reply:
you are right -- I meant `num_rows / 2` -- thank you

Reply:
As above, I think the `then` formulation is harder to read so I would like to leave the current less concise formulaton